libgalois/src/PropertyGraph.cpp

```diff
@@ -539,16 +647,44 @@ katana::Result<void>
 katana::PropertyGraph::DoWrite(
     tsuba::RDGHandle handle, const std::string& command_line,
     tsuba::RDG::RDGVersioningPolicy versioning_action) {
+  std::unique_ptr<tsuba::FileFrame> topology_res = nullptr;
+  std::unique_ptr<tsuba::FileFrame> node_entity_type_id_array_res = nullptr;
+  std::unique_ptr<tsuba::FileFrame> edge_entity_type_id_array_res = nullptr;
+
   if (!rdg_.topology_file_storage().Valid()) {
-    auto result = WriteTopology(topology());
-    if (!result) {
-      return result.error();
+    auto res = WriteTopology(topology());
+    if (!res) {
+      return res.error();
+    }
+    topology_res = std::move(res.value());
+  }
+
+  if (!rdg_.node_entity_type_id_array_file_storage().Valid()) {
+    auto res =
+        WriteEntityTypeIDsArray(GetNodeEntityTypeIDArrayData(), num_nodes());
+    if (!res) {
+      return res.error();
+    }
+    node_entity_type_id_array_res = std::move(res.value());
+  }
+
+  if (!rdg_.edge_entity_type_id_array_file_storage().Valid()) {
+    auto res =
+        WriteEntityTypeIDsArray(GetEdgeEntityTypeIDArrayData(), num_edges());
+    if (!res) {
+      return res.error();
     }
-    return rdg_.Store(
-        handle, command_line, versioning_action, std::move(result.value()));
+    edge_entity_type_id_array_res = std::move(res.value());
   }
 
-  return rdg_.Store(handle, command_line, versioning_action);
+  return rdg_.Store(
+      handle, command_line, versioning_action, std::move(topology_res),
+      std::move(node_entity_type_id_array_res),
+      std::move(edge_entity_type_id_array_res),
+      FlattenEntitySetOfIDsMap(node_entity_type_id_to_atomic_entity_type_ids_),
```

Comment:
This is wrong because `Flatten***` returns a range of iterators to a local variable, which will get destroyed when you return from the Flatten call. 

Flatten should instead return a vector of pairs, whose iterator range you can then pass to `rdg_.Store()`.